draft-ietf-ppm-dap.md

```diff
@@ -1544,7 +1544,7 @@ struct {
 
 struct {
   opaque agg_param<0..2^32-1>;
-  PartialBatchSelector part_batch_selector;
+  opaque part_batch_selector<0..2^16-1>;
```

Comment:
The justification for making this change is to allow unrecognized variants to be decoded, and to separate serialization logic from protocol logic.

However, I think this change doesn't meet those goals: before the change, we would fail to deserialize `AggregationJobInitReq` if there was an unknown batch mode variant; with the change, we will fail to deserialize `PartialBatchSelector` if there was an unknown batch mode variant.

I think the right way to address this would be to make the batch-mode-specific part of the message (i.e. the `select`) opaque instead, e.g:

```
struct {
  BatchMode batch_mode;
  opaque encoded_partial_batch_selector<0..2^16-1>;
} PartialBatchSelector;
```

With this change, we can decode the entire `PartialBatchSelector` message (or any message that contains it), then do application-level checks that the batch mode variant is known before we attempt to decode the batch-mode-specific data.

===

But stepping back a bit, given that the correct behavior when parsing an unknown batch mode is "fail immediately", I think introducing a new length field overcomplicates things. An implementation can arrange for its deserialization code to return a specific error code when encountering an unknown enum variant, which we could then use to drive the corresponding DAP behavior. Since the enum always comes before the values whose type is determined by the enum value, this will always work. Perhaps this does not separate serialization logic quite as cleanly from application logic, but I think it's worth it in the sense that we aren't spending an extra few bytes per message on the wire.

Reply:
> ```
> struct {
>   BatchMode batch_mode;
>   opaque encoded_partial_batch_selector<0..2^16-1>;
> } PartialBatchSelector;
> ```
> 
> With this change, we can decode the entire `PartialBatchSelector` message (or any message that contains it), then do application-level checks that the batch mode variant is known before we attempt to decode the batch-mode-specific data.

Actually, I think this accomplishes the same thing. It's just a matter of where we put the length prefix. If we put the length prefix before the batch mode, we know we have one more byte to parse for the batch mode, then the rest is specific to the batch mode; if we put the length before the batch mode, then we know the rest is specific to the batch mode.

The reason I spelled it this way is to be consistent with the extension points in taskprov for VDAF variants. We can take your approach if you strongly prefer it, but I think it might complicate the "Extending this Document" text slightly.

> But stepping back a bit, given that the correct behavior when parsing an unknown batch mode is "fail immediately", I think introducing a new length field overcomplicates things. An implementation can arrange for its deserialization code to return a specific error code when encountering an unknown enum variant, which we could then use to drive the corresponding DAP behavior. Since the enum always comes before the values whose type is determined by the enum value, this will always work. Perhaps this does not separate serialization logic quite as cleanly from application logic, but I think it's worth it in the sense that we aren't spending an extra few bytes per message on the wire.

I agree it's not strictly necessary. The thing I'm trying to accomplish is allowing implementations to separate parsing logic from protocol logic. With this PR, wether I recognize a batch mode is irrelevant to whether I can parse the AggregationJobInitReq.

Reply:
I was thinking of this over the weekend, and it occurred to me that we can spell this much nicer if we put the length tag first as you suggest. Will update the PR today or tomorrow.

Reply:
Done, please take another look at the commit message before going through the whole PR.